java-core/google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java

```diff
@@ -767,4 +794,40 @@ public String getClientLibToken() {
   public String getQuotaProjectId() {
     return quotaProjectId;
   }
+
+  /** Returns the resolved endpoint for the Service to connect to Google Cloud */
+  public String getResolvedEndpoint(String serviceName) {
```

Comment:
I think `host` in java-core is basically `endpoint` in Gax without the port, based on the value of [DEFAULT_HOST](https://github.com/googleapis/sdk-platform-java/blob/main/java-core/google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java#L85). So we should either rename this to `getResolvedHost` and remove the port concatenation in `formatEndpoint`, or return `host:port` if `host` is specified.
Can you please verify the format of `host`?

Reply:
I think the format of host is intended to be `scheme://host:port`. For gRPC, the channel is expecting something in format of `host:port` and httpjson is open to anything.

I see that the handwritten libraries handle these inside their own codebase:
1. Storage gRPC (strips the scheme and adds a port if missing): https://github.com/googleapis/java-storage/blob/2b5b27eac3279db815b36b252830d0905ade0665/google-cloud-storage/src/main/java/com/google/cloud/storage/GrpcStorageOptions.java#L163-L170
2. Storage HttpJson (sets it directly): https://github.com/googleapis/java-storage/blob/2b5b27eac3279db815b36b252830d0905ade0665/google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java#L143
3. Spanner (enforces that a port exists): https://github.com/googleapis/java-spanner/blob/21f5eba7d46c9011803accf0aaf87e6072066419/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java#L1469-L1471
4. Bigquery (sets it directly): https://github.com/googleapis/java-bigquery/blob/7a1bbd2d33e71644d83c42b71c67bac099882a4a/google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/spi/v2/HttpBigQueryRpc.java#L108

I'll update this to return in `host:port` format since that's the minimum required from both transports. The handwritten libraries already check to add in the port if it's missing.